cpp/include/raft/comms/std_comms.hpp

```diff
@@ -0,0 +1,417 @@
+  switch (datatype) {
+    case datatype_t::CHAR:
+      return sizeof(char);
+    case datatype_t::UINT8:
+      return sizeof(uint8_t);
+    case datatype_t::INT32:
+      return sizeof(int);
+    case datatype_t::UINT32:
+      return sizeof(unsigned int);
+    case datatype_t::INT64:
+      return sizeof(int64_t);
+    case datatype_t::UINT64:
+      return sizeof(uint64_t);
+    case datatype_t::FLOAT32:
+      return sizeof(float);
+    case datatype_t::FLOAT64:
+      return sizeof(double);
+  }
+}
+
+ncclDataType_t get_nccl_datatype(const datatype_t datatype) {
+  switch (datatype) {
+    case datatype_t::CHAR:
+      return ncclChar;
+    case datatype_t::UINT8:
+      return ncclUint8;
+    case datatype_t::INT32:
+      std::cout << "Returning int32" << std::endl;
+      return ncclInt;
+    case datatype_t::UINT32:
+      return ncclUint32;
+    case datatype_t::INT64:
+      return ncclInt64;
+    case datatype_t::UINT64:
+      return ncclUint64;
+    case datatype_t::FLOAT32:
+      return ncclFloat;
+    case datatype_t::FLOAT64:
+      return ncclDouble;
+  }
+}
+
+ncclRedOp_t get_nccl_op(const op_t op) {
+  switch (op) {
+    case op_t::SUM:
+      return ncclSum;
+    case op_t::PROD:
+      return ncclProd;
+    case op_t::MIN:
+      return ncclMin;
+    case op_t::MAX:
+      return ncclMax;
+  }
+}
+
+class std_comms : public comms_iface {
+ public:
+  std_comms() = delete;
+
+  /**
+   * @brief Constructor for collective + point-to-point operation.
+   * @param comm initialized nccl comm
+   * @param ucp_worker initialized ucp_worker instance
+   * @param eps shared pointer to array of ucp endpoints
+   * @param size size of the cluster
+   * @param rank rank of the current worker
+   */
+  std_comms(ncclComm_t nccl_comm, ucp_worker_h ucp_worker,
+            std::shared_ptr<ucp_ep_h *> eps, int num_ranks, int rank,
+            const std::shared_ptr<mr::device::allocator> device_allocator)
+    : nccl_comm_(nccl_comm),
+      ucp_worker_(ucp_worker),
+      ucp_eps_(eps),
+      num_ranks_(num_ranks),
+      rank_(rank),
+      device_allocator_(device_allocator),
+      next_request_id_(0) {
+    initialize();
+  };
+
+  /**
+   * @brief constructor for collective-only operation
+   * @param comm initilized nccl communicator
+   * @param size size of the cluster
+   * @param rank rank of the current worker
+   */
+  std_comms(const ncclComm_t nccl_comm, int num_ranks, int rank,
+            const std::shared_ptr<mr::device::allocator> device_allocator)
+    : nccl_comm_(nccl_comm),
+      num_ranks_(num_ranks),
+      rank_(rank),
+      device_allocator_(device_allocator) {
+    initialize();
+  };
+
+  virtual ~std_comms() {
+    CUDA_CHECK_NO_THROW(cudaStreamDestroy(stream_));
+
+    device_allocator_->deallocate(sendbuff_, sizeof(int), stream_);
+    device_allocator_->deallocate(recvbuff_, sizeof(int), stream_);
+  }
+
+  void initialize() {
+    CUDA_CHECK(cudaStreamCreate(&stream_));
```

Comment:
bit confused here. Why do we need a separate `stream_` member when all other functions are accepting `stream` parameter?